astacus/coordinator/coordinator.py

```diff
@@ -271,8 +271,9 @@ async def try_run(self, cluster: Cluster, context: StepsContext) -> bool:
                 with self._progress_handler(cluster, step):
                     try:
                         r = await step.run_step(cluster, context)
-                    except (StepFailedError, WaitResultError) as e:
-                        logger.info("Step %s failed: %s", step, str(e))
+                    except (StepFailedError, WaitResultError) as exc:
+                        logger.info("Step %s failed: %s", step, str(exc))
+                        await step.handle_step_failure(cluster, context)
```

Comment:
Oh I was thinking maybe more using a stack like

```python
failure_functions = []
for i, step in enumerate(self.steps, 1):
    failure_functions.append(step.handle_step_failure)
    ...
    try:
        ...
    except ...:
        for func in reversed(failure_functions):
            func(cluster, context)
```

But this is probably good enough

Reply:
Reading the code, if I understand correctly there are two scenarios:
1. If the failure is `Transient` then the step will be retried, so there's nothing to put on the stack yet. 
2. If the failure is permanent (`astacus.common.exceptions.PermanentException` subclass) then there won't be a next step. The exception clause in question sets the step to failed, and higher up the caller stack the whole `Op` is aborted (in this case the `BackupOp`).

Which case do you think could result in multiple steps needing to be unwound? 

> maybe the except needs to be more broad?
Looking at the usages for other `PermanentException` subclasses they're either raised during the restore steps or by the Rohmu storage subclasses (which would be caught when the server would be setup I think). Might be worth broadening `StepFailedError` to `PermanentException`..